Add Windows Compat Pack 2.0.0 to PSCore6 and update package references to final versions including dotnetcore 2.1

SteveL-MSFT opened the pull request:
## PR Summary

To support PowerShell Modules built with .Net Windows Compatibility Pack, we decided that it was best to ship the WCP assemblies with PSCore6.  This also adds many new .Net apis be default while only adding ~3.5 MB additional disk footprint (to a ~137 MB install currently).

Added ref to WCP 2.0.0 to SDK csproj.

Updated all refs of 4.5.0-RC1 to 4.5.0 final.

Update dotnetcore2.1 to final build

Update WebListener to aspnetcore.app 2.1.0 final
(Microsoft.VisualStudio.Web.CodeGeneration.Tools is not final yet, so no change)

## PR Checklist

- [x] [PR has a meaningful title](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---submission)
    - Use the present tense and imperative mood when describing your changes
- [x] [Summarized changes](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---submission)
- [x] [Change is not breaking](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#making-breaking-changes)
- [x] [Make sure all `.h`, `.cpp`, `.cs`, `.ps1` and `.psm1` files have the correct copyright header](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---submission)
- [x] This PR is ready to merge and is not [Work in Progress](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---work-in-progress).
    - If the PR is work in progress, please add the prefix `WIP:` to the beginning of the title and remove the prefix when the PR is ready.
- **User-facing changes**
    - [x] Not Applicable
    - **OR**
    - [ ] User-facing [Documentation needed](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---submission)
        - [ ] Issue filed - Issue link:
- **Testing - New and feature**
    - [x] Not Applicable or can only be tested interactively
    - **OR**
    - [ ] [Make sure you've added a new test if existing tests do not effectively test the code changed](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#before-submitting)
        - [ ] [Add `[feature]` if the change is significant or affects feature tests](https://github.com/PowerShell/PowerShell/blob/master/docs/testing-guidelines/testing-guidelines.md#requesting-additional-tests-for-a-pr)

Commit:
add Windows Compat Pack 2.0.0 to PSCore6

daxian-dbw commented on src/powershell-win-core/powershell-win-core.csproj:
```diff
@@ -38,6 +38,7 @@
     <PackageReference Include="PSDesiredStateConfiguration" Version="6.0.0-beta.8" />
     <PackageReference Include="PowerShellHelpFiles" Version="1.0.0-*" />
     <PackageReference Include="psrp.windows" Version="6.1.0-*" />
+    <PackageReference Include="Microsoft.Windows.Compatibility" Version="2.0.0" />
```
I think this reference should be put in `SDK.csproj` with a `Condition` attribute, like `<PackageReference Include="Microsoft.Windows.Compatibility" Version="2.0.0" Condition=" '$(IsWindows)' != 'true' " />`. In this way, application that refernce to `SDK` nuget package will get those extended pack too.

`IsWindows` is defined in https://github.com/PowerShell/PowerShell/blob/master/PowerShell.Common.props#L113

daxian-dbw replied:
/cc @adityapatwardhan, in case I'm wrong about the SDK reference part.

daxian-dbw replied:
Actually, I don't know if we should even have the `condition` part. Some of the compat assemblies work on Unix platforms too.

adityapatwardhan replied:
Yes, this should be part of `SDK`, so apps hosting PowerShell can use the modules with WCP.

SteveL-MSFT replied:
Perhaps we don't need the condition as @daxian-dbw is correct that some of the assemblies from the WCP meta-package work on non-Windows so we should make those available.  I'll make the change.

Commit:
moved ref of WCP to SDK
updated all refs of 4.5.0-RC1 to 4.5.0 final

daxian-dbw commented:
Ha, it seems dotnet core 2.1 is officially out: https://www.nuget.org/packages/Microsoft.NETCore.App/2.1.0
Shall we update our build to use the official 2.1?

Commit:
update dotnetcore to 2.1.300 RTM

markekraus commented:
@SteveL-MSFT Can you push a commit with `[feature]` to ensure that the version bump didn't cause any regressions?

bergmeister commented:
@SteveL-MSFT I believe you also need to update the installer and cleanup comments, see my commit [here](https://github.com/bergmeister/PowerShell/commit/189940143cbacceb5e99ded85a388eca81173ccd)

Commit:
[feature]
update cache names for CI
update files.wxs to reflect new versions

iSazonov commented:
We have a issue with `System.Text.Encoding.CodePages`. Maybe not still published.

iSazonov commented:
We could use Version="4.5.*" for packages.

iSazonov commented:
>we decided that it was best to ship the WCP assemblies with PSCore6

Can we use it directly - to access registry, perfcounters and so on?

SteveL-MSFT commented:
@iSazonov System.Text.Encoding.CodePages 4.5.0 is published: https://www.nuget.org/packages/System.Text.Encoding.CodePages/

Commit:
[feature]
update pwrshplugin.dll trusted assembly list

daxian-dbw commented on src/powershell-native/nativemsh/pwrshcommon/pwrshcommon.cpp:
```diff
@@ -669,9 +669,11 @@ namespace NativeMsh
-        "Microsoft.CodeAnalysis",
         "Microsoft.CodeAnalysis.CSharp",
+        "Microsoft.CodeAnalysis",
+        "Microsoft.CodeAnalysis.VisualBasic",
```
Do we want to include this one? I think we will remove VB support before 6.1 RC. If we include it here now, we will have to update the windows.psrp package again when removing the VB support.

SteveL-MSFT replied:
Will remove.  We can always add it back later if needed.

daxian-dbw commented on src/Microsoft.PowerShell.SDK/Microsoft.PowerShell.SDK.csproj:
```diff
@@ -15,20 +15,21 @@
   </ItemGroup>
 
   <ItemGroup>
+    <PackageReference Include="Microsoft.Windows.Compatibility" Version="2.0.0" />
```
If we don't know the source location of this package, maybe put it at the end, in the section with the comment ` the source could not be found for the following package(s)`, along with `Microsoft.NETCore.Windows.ApiSets`

SteveL-MSFT replied:
Ok

Commit:
[feature]
address Dongbo's feedback

SteveL-MSFT commented:
Can't tell why System.Text.Encoding.CodePages is failing to load as the signature matches the assembly.

iSazonov commented:
I suspect this is a publishing error. Could you ask .Net team?

SteveL-MSFT commented:
@iSazonov that xunit test passes fine on my desktop machine, just not in AppVeyor though

daxian-dbw commented:
I see the following from `Start-PSxUnit`:
```powershell
            $requiredDependencies = @(
                $nativeLib,
                "$Content/Microsoft.Management.Infrastructure.dll",
                "$Content/System.Text.Encoding.CodePages.dll"
            )
```
The `codepages.dll` line is suspicious to me. But I haven't spent time digging further yet.

Commit:
[feature]
Removed unncessary dep ref for psxunit on System.Text.Encoding.CodePages.dll
Updated packaging nupkg dep assembly versions

iSazonov commented:
Errors in Appveyor tests:
```
[00:21:47]   Describing CmsMessage cmdlets thorough tests
[00:21:48] +++++++++++++++++++?+
[00:21:48] 
[00:21:48] New-Item : The directory specified, 'Modules', is not a subdirectory of 'C:\'.
[00:21:48] Parameter name: path
[00:21:48] At C:\projects\powershell\test\powershell\Modules\Microsoft.PowerShell.Security\ConstrainedLanguageRestriction.Tests.ps1:87 char:21
[00:21:48] + ...          $null = New-Item -ItemType Directory $moduleDirectory -Force
[00:21:48] +                      ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
[00:21:48] + CategoryInfo          : InvalidArgument: (C:\Modules:String) [New-Item], ArgumentException
[00:21:48] + FullyQualifiedErrorId : CreateDirectoryArgumentError,Microsoft.PowerShell.Commands.NewItemCommand
```

```
[00:22:22]   Describing Invoke-Item basic tests
[00:22:22] 
[00:22:22]     Context Invoke a text file on Unix
[00:22:22] !!
[00:22:22] Get-Process : Cannot find a process with the name "notepad". Verify the process name and call the cmdlet again.
[00:22:22] At C:\projects\powershell\test\powershell\Modules\Microsoft.PowerShell.Utility\Invoke-Item.Tests.ps1:70 char:17
[00:22:22] +                 Get-Process -Name $notepadProcessName | Stop-Process  ...
[00:22:22] +                 ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
[00:22:22] + CategoryInfo          : ObjectNotFound: (notepad:String) [Get-Process], ProcessCommandException
[00:22:22] + FullyQualifiedErrorId : NoProcessFoundForGivenName,Microsoft.PowerShell.Commands.GetProcessCommand
```

```
[00:37:57] new file {$(env.ProductSourcePath)\Microsoft.Win32.SystemEvents.dll} need to be added to {C:\projects\powershell\tools\packaging\..\..\assets\Files.wxs}
[00:37:57] new file {$(env.ProductSourcePath)\System.CodeDom.dll} need to be added to {C:\projects\powershell\tools\packaging\..\..\assets\Files.wxs}
[00:37:57] new file {$(env.ProductSourcePath)\System.ComponentModel.Composition.dll} need to be added to {C:\projects\powershell\tools\packaging\..\..\assets\Files.wxs}
[00:37:57] new file {$(env.ProductSourcePath)\System.Configuration.ConfigurationManager.dll} need to be added to {C:\projects\powershell\tools\packaging\..\..\assets\Files.wxs}
[00:37:57] new file {$(env.ProductSourcePath)\System.Data.DataSetExtensions.dll} need to be added to {C:\projects\powershell\tools\packaging\..\..\assets\Files.wxs}
[00:37:57] new file {$(env.ProductSourcePath)\System.Data.Odbc.dll} need to be added to {C:\projects\powershell\tools\packaging\..\..\assets\Files.wxs}
[00:37:57] new file {$(env.ProductSourcePath)\System.Diagnostics.PerformanceCounter.dll} need to be added to {C:\projects\powershell\tools\packaging\..\..\assets\Files.wxs}
[00:37:57] new file {$(env.ProductSourcePath)\System.DirectoryServices.AccountManagement.dll} need to be added to {C:\projects\powershell\tools\packaging\..\..\assets\Files.wxs}
[00:37:57] new file {$(env.ProductSourcePath)\System.DirectoryServices.dll} need to be added to {C:\projects\powershell\tools\packaging\..\..\assets\Files.wxs}
[00:37:57] new file {$(env.ProductSourcePath)\System.DirectoryServices.Protocols.dll} need to be added to {C:\projects\powershell\tools\packaging\..\..\assets\Files.wxs}
[00:37:57] new file {$(env.ProductSourcePath)\System.Drawing.Common.dll} need to be added to {C:\projects\powershell\tools\packaging\..\..\assets\Files.wxs}
[00:37:57] new file {$(env.ProductSourcePath)\System.IO.Ports.dll} need to be added to {C:\projects\powershell\tools\packaging\..\..\assets\Files.wxs}
[00:37:57] new file {$(env.ProductSourcePath)\System.Management.dll} need to be added to {C:\projects\powershell\tools\packaging\..\..\assets\Files.wxs}
[00:37:57] new file {$(env.ProductSourcePath)\System.Runtime.Caching.dll} need to be added to {C:\projects\powershell\tools\packaging\..\..\assets\Files.wxs}
[00:37:57] new file {$(env.ProductSourcePath)\System.Security.Cryptography.ProtectedData.dll} need to be added to {C:\projects\powershell\tools\packaging\..\..\assets\Files.wxs}
[00:37:57] new file {$(env.ProductSourcePath)\System.Security.Cryptography.Xml.dll} need to be added to {C:\projects\powershell\tools\packaging\..\..\assets\Files.wxs}
[00:37:57] new file {$(env.ProductSourcePath)\System.ServiceModel.Syndication.dll} need to be added to {C:\projects\powershell\tools\packaging\..\..\assets\Files.wxs}
[00:37:57] 
```

SteveL-MSFT commented:
Ok, figured out the problem with System.Text.Encoding.CodePages.dll.  The nupkg doesn't contain unix runtime only Windows.  Older version has both.

Commit:
[feature]
revert ref to System.Text.Encoding.CodePages to 4.3.0 as newer versions don't have Unix runtime
updated files.wxs

Commit:
[feature]
revert CodePages back to 4.5.0-rc1

daxian-dbw committed:
[Feature] Update xunit packages

daxian-dbw commented:
PSXunit tests passed on my machine (consistently failing previously) after I update the xunit packages. So push a change to see if that also fixes the issue in CI.

SteveL-MSFT commented:
AppVeyor failure is due to recent PR to migrate the install docs, the aka.ms link now returned 404.  I updated the aka.ms link so restarting AppVeyor.

daxian-dbw reviewed (approved):
LGTM

iSazonov commented on src/System.Management.Automation/System.Management.Automation.csproj:
```diff
@@ -11,12 +11,12 @@
+    <PackageReference Include="System.Text.Encoding.CodePages" Version="4.5.0" />
     <!-- the following package(s) are from the powershell org -->
     <PackageReference Include="Microsoft.Management.Infrastructure" Version="1.0.0-alpha*" />
     <PackageReference Include="PowerShell.Core.Instrumentation" Version="6.0.0-beta*" />
```
Have we this APIs in WCP?

SteveL-MSFT replied:
These are pkgs PowerShell Team publishes

iSazonov replied:
Should we migrate to WCP assemplies if we have them on long time?

SteveL-MSFT replied:
You mean include them in WCP?  Probably not.

iSazonov commented on src/powershell-native/nativemsh/pwrshcommon/pwrshcommon.cpp:
```diff
@@ -669,9 +669,10 @@ namespace NativeMsh
         "Microsoft.ApplicationInsights",
-        "Microsoft.CodeAnalysis",
         "Microsoft.CodeAnalysis.CSharp",
+        "Microsoft.CodeAnalysis",
```
Why was the order changed?

SteveL-MSFT replied:
The list was generated rather than trying to insert a large number of new assemblies into the list.  `Sort-Object` seems to prefer the period over end of string.

iSazonov commented on test/csharp/csharp.tests.csproj:
```diff
@@ -21,10 +21,10 @@
     <!-- DotNetCliToolReference element specifies the CLI tool that the user wants to restore in the context of the project. -->
-    <DotNetCliToolReference Include="dotnet-xunit" Version="2.3.1" />
-    <PackageReference Include="Xunit.SkippableFact" Version="1.3.3" />
+    <DotNetCliToolReference Include="dotnet-xunit" Version="2.4.0-beta.1.build3958" />
```
I think we should open new tracking issue to update the beta to release.
And Microsoft.VisualStudio.Web.CodeGeneration.Tools  too.

SteveL-MSFT replied:
Created https://github.com/PowerShell/PowerShell/issues/6977

iSazonov commented on tools/packaging/packaging.psm1:
```diff
@@ -1409,12 +1409,12 @@ function New-UnifiedNugetPackage
                 'Microsoft.PowerShell.Commands.Management' {
                     $deps.Add([tuple]::Create([tuple]::Create('id', 'Microsoft.PowerShell.Security'), [tuple]::Create('version', $PackageVersion))) > $null
-                    $deps.Add([tuple]::Create([tuple]::Create('id', 'System.ServiceProcess.ServiceController'), [tuple]::Create('version', '4.4.1'))) > $null
+                    $deps.Add([tuple]::Create([tuple]::Create('id', 'System.ServiceProcess.ServiceController'), [tuple]::Create('version', '4.5.0'))) > $null
```
Not related the PR - why we set the versions manually and do not take from csproj files?

SteveL-MSFT replied:
https://github.com/PowerShell/PowerShell/issues/6978

iSazonov commented on assets/files.wxs:
```diff
@@ -4,6 +4,57 @@
         <DirectoryRef Id="$(var.ProductDirectoryName)">
+            <Component Id="cmpCBE3857586454EB3B634B118E8EBD1D1" Guid="{CBE38575-8645-4EB3-B634-B118E8EBD1D1}">
+                <File Id="filCBE3857586454EB3B634B118E8EBD1D1" KeyPath="yes" Source="$(env.ProductSourcePath)\Microsoft.Win32.SystemEvents.dll" />
+            </Component>
```
I lost understanding - why do we update the file manually? Can we generate it automatically?

SteveL-MSFT replied:
https://github.com/PowerShell/PowerShell/issues/6978

iSazonov reviewed (approved):
Leave a comment

iSazonov commented:
Newtonsoft.Json latest version is 11.0.1
https://github.com/JamesNK/Newtonsoft.Json/releases

NJsonSchema latest version is 9.10.50
https://www.nuget.org/packages/NJsonSchema/

Commit:
[feature]
update Newtonsoft.Json and NJsonSchema to latest versions
